perf(wasm): generate scope text with visitor

overlookmotel opened the pull request:
Follow-on after #5232. `oxc_wasm` build scopes text with a single AST traversal. Previous implementation was O($n^2$).

If we can assume scopes are listed in traversal order, then we could do it a bit more efficiently just from `ScopeTree`, but this approach of using `Visit` will handle out-of-order scope IDs (which you'd get if printing a post-transform `ScopeTree`).

Also reduce creating and discarding `String`s for indentation - reuse a single string instead.

overlookmotel commented:
* **#5264** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5264?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Dunqing commented:
I am considering merging scope and symbol into one, based on 

https://github.com/oxc-project/oxc/blob/813e10ce507760df23dea67aef88de6791b67da5/crates/oxc_semantic/tests/main.rs#L9-L96

Because I think keeping the consistent with our semantic snapshot is easy to debug the bugs

overlookmotel commented:
I see your point of it being easier for debugging if the format is consistent. But I think visually it might be a bit too much information all in one place to combine the 2. It'd work if you had expand/contract buttons so you didn't have to see everything at once. Maybe that's achievable with our new fancy Playground?

overlookmotel committed:
perf(wasm): generate scope text with visitor (#5264)

Follow-on after #5232. `oxc_wasm` build scopes text with a single AST traversal. Previous implementation was O($n^2$).

If we can assume scopes are listed in traversal order, then we could do it a bit more efficiently just from `ScopeTree`, but this approach of using `Visit` will handle out-of-order scope IDs (which you'd get if printing a post-transform `ScopeTree`).

Also reduce creating and discarding `String`s for indentation - reuse a single string instead.
